File: Computer_Science/Algorithm/Learning/Algorithms/回溯/组合型/code/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

#include <string>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
    
    string path = "";
    vector<string> ans;
    
    // 我们现在遍历到可选括号数组中的第i个位置
    void dfs(int n, int i, int open) {
        if (i == n) {
            ans.emplace_back(path);
            return;
        }
        
        // 选左括号< 也就是现在左括号的数量 < n / 2的时候
        if (open < n / 2) {
            path += '(';
            dfs(n, i+1, open+1);
            path.pop_back();
        }
        // 选右括号, 也就是现在还有左括号能匹配, 也就是path中右括号的数量 < 左括号的数量
        if ((i - open) < open) {
            path += ')';
            dfs(n, i+1, open);
            path.pop_back();
        }
    }

    vector<string> generateParenthesis(int n) {
        dfs(2*n, 0, 0);
        return ans;
    }
};
// @lc code=end

File: Computer_Science/Algorithm/leetcode/hot100/22.括号生成.java
/*
 * @lc app=leetcode.cn id=22 lang=java
 *
 * [22] 括号生成
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        char[] path = new char[2*n];
        dfs(n, n, ans, path, 0);
        return ans;
    }

    /**
     * 全排列问题, 这里是有两种选择, 但是只能在rightNum > leftNum的时候选右括号
     * @param leftNum
     * @param rightNum
     * @param ans
     * @param path
     * @param i
     */
    private void dfs(int leftNum, int rightNum, List<String> ans, char[] path, int i) {
        if (leftNum == 0 && rightNum == 0) {
            ans.add(new String(path));
            return;
        }

        // 选左括号
        if (leftNum > 0) {
            path[i] = '(';
            dfs(leftNum-1, rightNum, ans, path, i+1);
        }

        if (rightNum > 0 && rightNum > leftNum) {
            path[i] = ')';
            dfs(leftNum, rightNum-1, ans, path, i+1);
        }
    }
}
// @lc code=end

